### PR TITLE
backstage: use paths filter to find components to check for percy runs

### PR DESCRIPTION
**Pull request by frshwtr**
## Checklist before requesting a review
- [ ] I have applied `percy` label on my PR before merging and after review.
- [ ] If it is a new feature, I have added thorough tests.
- [ ] I have updated relevant docs.

### CODE REVIEW

File: templates/percy-workflow.yml [hunk 1]
```diff
@@ -25,26 +25,55 @@ on:
...
+          volta-version: "1.0.8"
+          path: ~/.npm
+          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
+      - run: npm ci
+      - name: Find components affected by changes
+        id: percy-changes
+        run: |
+          git fetch
```
Comment (frshwtr) [inline templates/percy-workflow.yml:48]: This shouldn't be needed, our checkout command above should be able to do it.

File: templates/percy-workflow.yml [hunk 2]
```diff
@@ -25,26 +25,55 @@ on:
...
+          git branch
```
Comment (frshwtr) [inline templates/percy-workflow.yml:49]: Remove this

File: templates/percy-workflow.yml [hunk 3]
```diff
@@ -25,26 +25,55 @@ on:
...
+          changes=()
+          for component in components/*/; do
+            RESULT=$(WORKSPACE=$component node ./scripts/should-percy-run.js | grep "\-RESULT:" | cut -d ":" -f 2)
+            echo "RESULT=$RESULT"
```
Comment (frshwtr) [inline templates/percy-workflow.yml:53]: Not needed

File: templates/percy-workflow.yml [hunk 4]
```diff
@@ -25,26 +25,55 @@ on:
...
+            if [[ $RESULT = 'true' ]]; then
+              echo "Adding $component to changes"
```
Comment (frshwtr) [inline templates/percy-workflow.yml:55]: This could be more descripted, why is it being added?

File: templates/percy-workflow.yml [hunk 5]
```diff
@@ -25,26 +25,55 @@ on:
...
+              changes+=("$component")
+            fi
+          done
+          echo "Changes:";
+          json_array=$(jq -n '$ARGS.positional' --args "${changes[@]}");
```
Comment (frshwtr) [inline templates/percy-workflow.yml:60]: Re order here so it's clear what changes are being logged

File: templates/percy-workflow.yml [hunk 6]
```diff
@@ -1,50 +1,95 @@
...
+        - uses: actions/checkout@v2
+        - uses: volta-cli/action@v4
+        - uses: actions/cache@v2
+          with:
+            volta-version: "1.0.8"
+            path: ~/.npm
+            key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+            restore-keys: |
+              ${{ runner.os }}-node-
+        - run: npm ci
```
Comment (frshwtr) [inline templates/percy-workflow.yml:87]: See if you can run these once for the whole matrix, to save on time

File: templates/percy-workflow.yml [hunk 7]
```diff
@@ -1,50 +1,83 @@
...
+            fetch-depth: 0
+        - uses: volta-cli/action@v4
+        - uses: actions/cache@v2
+          with:
+            volta-version: "1.0.8"
+            path: ~/.npm
+            key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+            restore-keys: |
+              ${{ runner.os }}-node-
+        - run: npm ci
```
Comment (frshwtr) [inline templates/percy-workflow.yml:46]: This isn't going to work, you need to restore a cache of the cached files before using it in test

File: templates/percy-workflow.yml [hunk 8]
```diff
@@ -25,26 +25,71 @@ on:
...
-    - run: node ./scripts/percy.js
-      env:
-        PERCY_TOKEN: ${{ secrets.PERCY_TOKEN_<% percyTokenName %> }}
-        WORKSPACE: <%& workspace %>
-  <%/projects%>
+          for component in components/*/; do
+            RESULT=$(WORKSPACE=$component node ./scripts/should-percy-run.js | grep "\-RESULT:" | cut -d ":" -f 2)
+            component_trimmed=$(echo $component | sed 's/\/$//g')
+            if [[ $RESULT = 'true' ]]; then
+              echo "$component_trimmed is affected changes, adding to percy run list."
```
Comment (frshwtr) [inline templates/percy-workflow.yml:56]: ```...is affected by changes...`

File: templates/percy-workflow.yml [hunk 9]
```diff
@@ -25,26 +25,71 @@ on:
...
+          restore-keys: |
+            ${{ runner.os }}-node-
+      - run: npm ci
+      - name: Start Percy check
+        env:
+          PERCY_TOKEN: ${{ secrets[matrix.token_name] }}
+          WORKSPACE: ${{ matrix.component }}
+        run: |
+          echo "Running percy for ${{ matrix.component }}"
+          echo "Search for component: ${{ matrix.token_name }}"
```
Comment (frshwtr) [inline templates/percy-workflow.yml:94]: Remove this line

File: scripts/lib/should-percy-run.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+		$.verbose = true
+		let messages = commits.stdout.split('\n');
+		for (const message of messages) {
+			if (message.startsWith('fix:') || message.startsWith('fix!:') || message.startsWith('feat:') || message.startsWith('feat!:')) {
+				$.verbose = false
+				let {stdout: changedFiles} = await $`git diff --name-only origin/${baseRef}...origin/${headRef}`;
+				$.verbose = true
+				const arrayOfChangeFiles = changedFiles.split('\n')
+				const effectingFiles = changedFileEffectsPercy(arrayOfChangeFiles)
+				console.log(`🚀 ~ changedFiles:`, arrayOfChangeFiles, effectingFiles);
```
Comment (akomiqaia) [inline scripts/lib/should-percy-run.js:35]:
```suggestion
```

File: scripts/lib/should-percy-run.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import {join} from "node:path";
+import {readPackage} from "read-pkg";
+import * as workspaces from "./workspaces.js";
+
+const isPullRequest = context.payload.pull_request
+
+export async function shouldPercyRun(dirname, workspace) {
+	const isDefaultBranch = context.ref.endsWith("/main")
+	const isChoreRelease = isPullRequest && context.payload.pull_request.title == 'chore: release main'
+	console.log(`🚀 ~ context.ref:`, context.ref, {isChoreRelease});
```
Comment (akomiqaia) [inline scripts/lib/should-percy-run.js:13]:
```suggestion
```

File: templates/percy-workflow.yml (hunk 8 above)
Comment (notlee) [reply templates/percy-workflow.yml:56]: I wonder if this bash could be simplified if the results were returns as JSON, which we could interpret with jq? My brain broke a little reviewing 😄 However in the interest of focusing on our priority, I'm happy to approve this PR if you've been able to manually test thoroughly?

**Review by akomiqaia — COMMENTED** [review]
I think it was me who left this logs before and not sure if we needed anymore.

**Pull request merged**